presto-base-arrow-flight/src/main/java/com/facebook/plugin/arrow/ArrowPageSource.java

```diff
@@ -0,0 +1,163 @@
+import com.facebook.presto.spi.ConnectorSession;
+import org.apache.arrow.flight.FlightRuntimeException;
+import org.apache.arrow.flight.FlightStream;
+import org.apache.arrow.flight.Ticket;
+import org.apache.arrow.vector.FieldVector;
+import org.apache.arrow.vector.VectorSchemaRoot;
+import org.apache.arrow.vector.dictionary.Dictionary;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import static com.facebook.plugin.arrow.ArrowErrorCode.ARROW_FLIGHT_ERROR;
+
+public class ArrowPageSource
+        implements ConnectorPageSource
+{
+    private static final Logger logger = Logger.get(ArrowPageSource.class);
+    private final ArrowSplit split;
+    private final List<ArrowColumnHandle> columnHandles;
+    private boolean completed;
+    private int currentPosition;
+    private Optional<VectorSchemaRoot> vectorSchemaRoot = Optional.empty();
+    private ArrowFlightClient flightClient;
+    private FlightStream flightStream;
+
+    public ArrowPageSource(ArrowSplit split, List<ArrowColumnHandle> columnHandles, ArrowFlightClientHandler clientHandler,
+            ConnectorSession connectorSession)
+    {
+        this.columnHandles = columnHandles;
+        this.split = split;
+        getFlightStream(clientHandler, split.getTicket(), connectorSession);
+    }
+
+    private void getFlightStream(ArrowFlightClientHandler clientHandler, byte[] ticket, ConnectorSession connectorSession)
+    {
+        try {
+            Optional<String> uri = (split.getLocationUrls().isEmpty()) ?
+                    Optional.empty() : Optional.of(split.getLocationUrls().get(0));
+            flightClient = clientHandler.getClient(uri);
+            flightStream = flightClient.getFlightClient().getStream(new Ticket(ticket), clientHandler.getCallOptions(connectorSession));
+        }
+        catch (FlightRuntimeException e) {
+            throw new ArrowException(ARROW_FLIGHT_ERROR, e.getMessage(), e);
+        }
+    }
+
+    @Override
+    public long getCompletedBytes()
+    {
+        return 0;
+    }
+
+    @Override
+    public long getCompletedPositions()
+    {
+        return currentPosition;
+    }
+
+    @Override
+    public long getReadTimeNanos()
+    {
+        return 0;
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return completed;
+    }
+
+    @Override
+    public long getSystemMemoryUsage()
+    {
+        return 0;
+    }
+
+    @Override
+    public Page getNextPage()
+    {
+        if (vectorSchemaRoot.isPresent()) {
+            vectorSchemaRoot.get().close();
+            vectorSchemaRoot = Optional.empty();
+        }
+
+        if (flightStream.next()) {
+            vectorSchemaRoot = Optional.ofNullable(flightStream.getRoot());
+        }
+
+        if (!vectorSchemaRoot.isPresent()) {
+            completed = true;
+        }
+
+        if (isFinished()) {
+            return null;
+        }
+
+        currentPosition++;
```

Comment:
```suggestion
        currentPosition = currentPosition + 1;
```

We tend to avoid `++` except in loops.

Reply:
Updated